File: src/main/java/ru/ioleksiv/telegram/bot/core/controller/handler/check/impl/ExtractValidator.java
package ru.ioleksiv.telegram.bot.core.controller.handler.check.impl;

import ru.ioleksiv.telegram.bot.core.controller.handler.check.Validator;
import ru.ioleksiv.telegram.bot.core.controller.handler.unpack.Unpacker;

import java.util.Optional;

public class ExtractValidator<IN, OUT> implements Validator<IN> {

    private final Unpacker<IN, OUT> unpacker;
    private final Validator<OUT> validator;

    public ExtractValidator(Unpacker<IN, OUT> unpacker,
                            Validator<OUT> validator) {
        this.unpacker = unpacker;
        this.validator = validator;
    }

    @Override
    public boolean check(IN argument) {
        return Optional.of(argument)
                .flatMap(unpacker::unpack)
                .filter(validator::check)
                .isPresent();

    }

}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/handler/check/impl/ExtractValidator.kt
package io.github.oleksivio.telegram.bot.core.controller.handler.check.impl

import io.github.oleksivio.telegram.bot.core.controller.handler.check.Validator
import io.github.oleksivio.telegram.bot.core.controller.handler.unpack.UnpackerFunction

class ExtractValidator<IN, OUT>(private val unpackerFunctionFunc: UnpackerFunction<IN, OUT>,
                                private val validator: Validator<OUT>) : Validator<IN> {
    override fun invoke(checkedValue: IN): Boolean {
        val unpacked = unpackerFunctionFunc(checkedValue) ?: return false
        return validator.invoke(unpacked)
    }

}
